timesketch/api/v1/resources/sketch.py

```diff
@@ -65,28 +65,86 @@ def get(self):
         Returns:
             List of sketches (instance of flask.wrappers.Response)
         """
+        args = self.parser.parse_args()
+        scope = args.get('scope', None)
+        page = args.get('page', 1)
+        search_query = args.get('search_query', None)
+
         if current_user.admin:
             sketch_query = Sketch.query
         else:
             sketch_query = Sketch.all_with_acl()
 
-        filtered_sketches = sketch_query.filter(
+        base_filter = sketch_query.filter(
+            not_(Sketch.Status.status == 'deleted'),
+            not_(Sketch.Status.status == 'archived'),
+            Sketch.Status.parent).order_by(Sketch.updated_at.desc())
+
+        base_filter_with_archived = sketch_query.filter(
             not_(Sketch.Status.status == 'deleted'),
-            Sketch.Status.parent).order_by(Sketch.updated_at.desc()).all()
+            Sketch.Status.parent).order_by(Sketch.updated_at.desc())
 
-        # Just return a subset of the sketch objects to reduce the amount of
-        # data returned.
+        filtered_sketches = base_filter_with_archived
         sketches = []
-        for sketch in filtered_sketches:
-            sketches.append({
+        return_sketches = []
+        num_hits = 0
+
+        if scope == 'recent':
+            views = View.query.filter_by(
+                user=current_user, name='').order_by(
+                View.updated_at.desc()).limit(3)
+            sketches = [view.sketch for view in views]
+            num_hits = len(sketches)
+        elif scope == 'user':
+            filtered_sketches = base_filter.filter_by(user=current_user)
+        elif scope == 'archived':
+            filtered_sketches = sketch_query.filter(
+                Sketch.status.any(status='archived'))
+        elif scope == 'shared':
+            filtered_sketches = base_filter.filter(Sketch.user != current_user)
+        elif scope == 'search':
+            filtered_sketches = base_filter_with_archived.filter(
+                or_(
+                    Sketch.name.ilike(f'%{search_query}%'),
+                    Sketch.description.ilike(f'%{search_query}%')
+                )
+            )
+
+        # If no scope is set, fall back to returning all sketches.
+        if not scope:
+            sketches = filtered_sketches.all()
+
+        if not sketches:
+            pagination = filtered_sketches.paginate(page=page, per_page=20)
+            sketches = pagination.items
+            num_hits = pagination.total
+
+        for sketch in sketches:
+            # Last time a user did a query in the sketch, indicating activity.
+            try:
+                last_activity = View.query.filter_by(
+                    sketch=sketch, name='').order_by(
+                    View.updated_at.desc()).first().updated_at
```

Comment:
and again, this is not a good indication of a last activity of a sketch... this only tells a very partial story... and if we woudl implement what I talked about in the other comment, then we would have a much more accurate picture, and you could just take the last updated timestamp of the sketch and make use of that...

Reply:
Yes, I agree. In a followup PR though.